prototype2/multigrid/mgmesytec.cpp

```diff
@@ -136,80 +120,39 @@ void CSPEC::input_thread() {
   cspecdata.setbuffers(EFUSettings.DetectorTxBufferSize, EFUSettings.DetectorRxBufferSize);
   cspecdata.printbuffers();
   cspecdata.settimeout(0, 100000); // One tenth of a second
-
-  int rdsize;
-  for (;;) {
-    unsigned int eth_index = eth_ringbuf->getindex();
-
-    /** this is the processing step */
-    eth_ringbuf->setdatalength(eth_index, 0);
-    if ((rdsize = cspecdata.receive(eth_ringbuf->getdatabuffer(eth_index),
-                                    eth_ringbuf->getmaxbufsize())) > 0) {
-      eth_ringbuf->setdatalength(eth_index, rdsize);
-      mystats.rx_packets++;
-      mystats.rx_bytes += rdsize;
-      XTRACE(INPUT, DEB, "rdsize: %u\n", rdsize);
-
-      if (input2proc_fifo.push(eth_index) == false) {
-        mystats.fifo_push_errors++;
-      } else {
-        eth_ringbuf->nextbuffer();
-      }
-    }
-
-    // Checking for exit
-    if (not runThreads) {
-      XTRACE(INPUT, ALW, "Stopping input thread.\n");
-      return;
-    }
-  }
-}
-
-void CSPEC::processing_thread() {
   Producer producer(EFUSettings.KafkaBroker, "C-SPEC_detector");
   Producer monitorprod(EFUSettings.KafkaBroker, "C-SPEC_monitor");
   FBSerializer flatbuffer(kafka_buffer_size, producer);
   ReadoutSerializer readouts(100000, monitorprod);
   HistSerializer histfb;
   NMXHists hists;
 
-  MesytecData dat;
+  bool dumptofile = !DetectorSettings.fileprefix.empty();
+  MesytecData dat(dumptofile, DetectorSettings.fileprefix, DetectorSettings.module);
 
   dat.setWireThreshold(DetectorSettings.wireThresholdLo, DetectorSettings.wireThresholdHi);
   dat.setGridThreshold(DetectorSettings.gridThresholdLo, DetectorSettings.gridThresholdHi);
 
+  char buffer[9010];
+  int rdsize;
   TSCTimer report_timer;
+  for (;;) {
+    if ((rdsize = cspecdata.receive(buffer, eth_buffer_size)) > 0) {
+      mystats.rx_packets++;
+      mystats.rx_bytes += rdsize;
+      XTRACE(INPUT, DEB, "rdsize: %u\n", rdsize);
 
-  unsigned int data_index;
-  while (1) {
-
-    if ((input2proc_fifo.pop(data_index)) == false) {
-      mystats.rx_idle1++;
-      usleep(1);
-    } else {
-      auto len = eth_ringbuf->getdatalength(data_index);
-      if (len == 0) {
-        mystats.fifo_seq_errors++;
-      } else {
-        auto res = dat.parse(eth_ringbuf->getdatabuffer(data_index),
-                  eth_ringbuf->getdatalength(data_index), hists, readouts);
-
-        if (res < 0) {
-          continue;
-        }
-
-        int pixel = dat.getPixel();
-        int time  = dat.getTime();
-        XTRACE(PROCESS, DEB, "Time %d, pixel: %d\n", time, pixel);
-        if (pixel != 0) {
-          mystats.tx_bytes += flatbuffer.addevent(time, pixel);
-          mystats.rx_events++;
-          mystats.rx_readouts += dat.readouts;
-          mystats.rx_discards += dat.discards;
-        } else {
-          mystats.geometry_errors++;
-        }
+      auto res = dat.parse(buffer, rdsize, hists, flatbuffer, readouts);
+      if (res < 0) {
```

Comment:
This file (`mgmesytec.cpp`) implements enough functionality that unit testing this code seems appropriate to me.